plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/ProjectsProvider.java

```diff
@@ -0,0 +1,179 @@
+ */
+
+package com.google.cloud.tools.eclipse.projectselector;
+
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.cloud.tools.eclipse.projectselector.model.GcpProject;
+import com.google.cloud.tools.eclipse.ui.util.DisplayExecutor;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Future;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.jface.viewers.IStructuredContentProvider;
+import org.eclipse.jface.viewers.Viewer;
+
+/**
+ * A JFace {@link IStructuredContentProvider} for retrieving projects for the {@link Credential}
+ * provided as its input.
+ */
+public class ProjectsProvider implements IStructuredContentProvider {
+  private static final Logger logger = Logger.getLogger(ProjectsProvider.class.getName());
+  private static final Object[] EMPTY_OBJECTS = new Object[0];
+
+  public interface Callback<T> {
+    void execute(T value);
+  }
+
+  private final ProjectRepository projectRepository;
+  private Executor displayExecutor;
+  private Viewer viewer;
+  private Credential credential; // the input
+  private FetchProjectsJob fetchProjectsJob;
+
+  public ProjectsProvider(ProjectRepository projectRepository) {
+    this.projectRepository = projectRepository;
+  }
+
+  @Override
+  public void dispose() {
+    cancel();
+  }
+
+  private ListenableFuture<GcpProject[]> query() {
+    if (fetchProjectsJob != null) {
+      fetchProjectsJob.cancel();
+    }
+    fetchProjectsJob = new FetchProjectsJob();
+    fetchProjectsJob.schedule();
+    return fetchProjectsJob.getProjects();
+  }
+
+  @Override
+  public void inputChanged(Viewer viewer, Object oldInput, Object newInput) {
+    this.viewer = viewer;
+    this.displayExecutor = DisplayExecutor.create(viewer.getControl().getDisplay());
+    if (newInput instanceof Credential && !viewer.getControl().isDisposed()) {
+      this.credential = (Credential) newInput;
+      query().addListener(new Runnable() {
+        @Override
+        public void run() {
+          if (!ProjectsProvider.this.viewer.getControl().isDisposed()) {
+            ProjectsProvider.this.viewer.refresh();
+          }
+        }
+      }, displayExecutor);
+    } else {
+      cancel();
+    }
+  }
+
+  @Override
+  public Object[] getElements(Object inputElement) {
+    if (fetchProjectsJob != null) {
+      Future<GcpProject[]> projectsFuture = fetchProjectsJob.getProjects();
+      if (projectsFuture != null && projectsFuture.isDone()) {
+        try {
+          return projectsFuture.get();
```

Comment:
One caveat: if you launch two jobs by switching from a credential A to a credential B quickly, you'll have two jobs running in parallel. It is possible that the earlier job (for the credential A) completes after the latest job (for the credential B). The result is that the account selector is on the credential B while the project selector lists projects from the credential A.

Reply:
The first job will be cancelled before the second job is scheduled. `checkCancelled()` should prevent the future from being set if either the job was explicitly cancelled or should the provider's credential changed (though now that I look at the code, the future should be cancelled too, and it should also support external cancelling the future).

Actually looking at the code for `FetchProjectsJob`, I should extract this out to a common base class for use with other implementations as there's a few gotchas.

Reply:
If you think about it, `checkCancelled()` cannot guard it; `Job.cancel()` may be called after the last location of `checkCancelled()` but before setting the future. In other words, `cancel()` does not guarantee cancellation, and the cancelled job may run to completion, triggering the listener. If timing is bad, the listener of the earlier job may execute later than the second job.

As you said, you also need to cancel the future, but since all `Future.cancel()` does is to set the cancelled flag (or additionally the thread interrupted flag), it could suffer from the same problem if not done correctly. Also note that `Future.cancel()` will make `Future.isDone()` return true.

Reply:
`checkCancelled()` also treats the `ProjectProvider`'s credential having changed as indicating cancellation.

Good point re: isDone() under cancelled;